v2/pkg/protocols/common/hosterrorscache/hosterrorscache_test.go

```diff
@@ -2,30 +2,112 @@ package hosterrorscache
 
 import (
 	"fmt"
+	"sync"
 	"testing"
 
 	"github.com/stretchr/testify/require"
 )
 
-func TestCacheCheckMarkFailed(t *testing.T) {
+func TestCacheCheck(t *testing.T) {
 	cache := New(3, DefaultMaxHostsCount)
 
-	cache.MarkFailed("http://example.com:80", fmt.Errorf("no address found for host"))
-	if value, err := cache.failedTargets.Get("http://example.com:80"); err == nil && value != nil {
-		require.Equal(t, 1, value, "could not get correct number of marked failed hosts")
-	}
-	cache.MarkFailed("example.com:80", fmt.Errorf("Client.Timeout exceeded while awaiting headers"))
-	if value, err := cache.failedTargets.Get("example.com:80"); err == nil && value != nil {
-		require.Equal(t, 2, value, "could not get correct number of marked failed hosts")
-	}
-	cache.MarkFailed("example.com", fmt.Errorf("could not resolve host"))
-	if value, err := cache.failedTargets.Get("example.com"); err == nil && value != nil {
-		require.Equal(t, 1, value, "could not get correct number of marked failed hosts")
-	}
 	for i := 0; i < 3; i++ {
 		cache.MarkFailed("test", fmt.Errorf("could not resolve host"))
 	}
 
 	value := cache.Check("test")
 	require.Equal(t, true, value, "could not get checked value")
 }
+
+func TestCacheMarkFailed(t *testing.T) {
+	cache := New(3, DefaultMaxHostsCount)
+
+	tests := []struct {
+		host     string
+		expected int
+	}{
+		{"http://example.com:80", 1},
+		{"example.com:80", 2},
+		{"example.com", 1},
+	}
+
+	for _, test := range tests {
+		normalizedCacheValue := cache.normalizeCacheValue(test.host)
+		cache.MarkFailed(test.host, fmt.Errorf("no address found for host"))
+		failedTarget, err := cache.failedTargets.Get(normalizedCacheValue)
+		require.Nil(t, err)
+		require.NotNil(t, failedTarget)
+
+		value, ok := failedTarget.(*cacheItem)
+		require.True(t, ok)
+		require.EqualValues(t, test.expected, value.errors.Load())
+	}
+}
+
+func TestCacheMarkFailedMultipleCalls(t *testing.T) {
+	cache := New(3, DefaultMaxHostsCount)
+
+	tests := []struct {
+		host     string
+		expected int
+	}{
+		{"http://example.com:80", 1},
+		{"example.com:80", 2},
+		{"example.com", 1},
+	}
+
+	for _, test := range tests {
+		normalizedCacheValue := cache.normalizeCacheValue(test.host)
+		cache.MarkFailed(test.host, fmt.Errorf("no address found for host"))
+		failedTarget, err := cache.failedTargets.Get(normalizedCacheValue)
+		require.Nil(t, err)
+		require.NotNil(t, failedTarget)
+
+		value, ok := failedTarget.(*cacheItem)
+		require.True(t, ok)
+		require.EqualValues(t, test.expected, value.errors.Load())
+
+		existingCacheItem, err := cache.failedTargets.GetIFPresent(normalizedCacheValue)
+		require.Nil(t, err)
+		require.NotNil(t, existingCacheItem)
+		existingCacheItemValue := existingCacheItem.(*cacheItem)
+		skippingValue := existingCacheItemValue.errors.Load() >= int32(test.expected)
+		require.Equal(t, true, skippingValue, "Didn't skipped host")
+	}
+}
+
+func TestCacheMarkFailedConcurrent(t *testing.T) {
```

Comment:
The test should look like the following one:
```go
func TestCacheMarkFailedConcurrent(t *testing.T) {
	cache := New(3, DefaultMaxHostsCount)

	tests := []struct {
		host     string
		expected int32
	}{
		{"http://example.com:80", 200},
		{"example.com:80", 200},
		{"example.com", 100},
	}

	wg := sync.WaitGroup{}
	for _, test := range tests {
		currentTest := test
		for i := 0; i < 100; i++ {
			wg.Add(1)

			go func() {
				defer wg.Done()

				cache.MarkFailed(currentTest.host, fmt.Errorf("could not resolve host"))
			}()
		}
	}

	wg.Wait()

	for _, test := range tests {
		require.True(t, cache.Check(test.host))

		normalizedCacheValue := cache.normalizeCacheValue(test.host)
		failedTarget, err := cache.failedTargets.Get(normalizedCacheValue)
		require.Nil(t, err)
		require.NotNil(t, failedTarget)

		value, ok := failedTarget.(*cacheItem)
		require.True(t, ok)
		require.EqualValues(t, test.expected, value.errors.Load())
	}
}
```
The purpose is to ensure that concurrent calls to `cache.MarkFailed` generate the final expected result (`cache.Check` true and corrent amount of errors. double for equal normalized values)